File: api-pytest-framework/agents/tracer.py
# ============================================================
# Tracer — Instrumentation des appels LLM
# ============================================================
# Enregistre chaque appel LLM dans logs/traces.jsonl.
# Format JSONL : une ligne JSON par appel, lisible par observability-agent.
#
# Utilisé par llm.py automatiquement — pas d'import manuel nécessaire.
# Chaque entrée contient :
#   ts, agent, fn, model, duration_ms, prompt_len, response_len,
#   success, confidence, error, retries
# ============================================================

import os, json, time, inspect, datetime

FRAMEWORK  = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
LOGS_DIR   = os.path.join(FRAMEWORK, "logs")
TRACE_FILE = os.path.join(LOGS_DIR, "traces.jsonl")

os.makedirs(LOGS_DIR, exist_ok=True)


def _detect_agent() -> str:
    """Remonte la stack pour trouver quel agent a appelé llm.*"""
    agents_dir = os.path.normpath(os.path.dirname(__file__))
    for frame_info in inspect.stack():
        fpath = os.path.normpath(frame_info.filename)
        if fpath.startswith(agents_dir) and "llm.py" not in fpath and "tracer.py" not in fpath:
            name = os.path.basename(fpath).replace(".py", "")
            return name
    return "unknown"


def record(
    fn: str,
    duration_ms: float,
    prompt_len: int,
    response_len: int,
    success: bool,
    model: str = "",
    confidence: float = None,
    error: str = None,
    retries: int = 0,
):
    """Écrit une entrée de trace dans logs/traces.jsonl."""
    entry = {
        "ts":           datetime.datetime.utcnow().isoformat() + "Z",
        "agent":        _detect_agent(),
        "fn":           fn,
        "model":        model,
        "duration_ms":  round(duration_ms, 1),
        "prompt_len":   prompt_len,
        "response_len": response_len,
        "success":      success,
        "retries":      retries,
    }
    if confidence is not None:
        entry["confidence"] = round(confidence, 3)
    if error:
        entry["error"] = str(error)[:200]

    with open(TRACE_FILE, "a", encoding="utf-8") as f:
        f.write(json.dumps(entry, ensure_ascii=False) + "\n")


class Span:
    """Context manager pour tracer un bloc d'appel LLM."""

    def __init__(self, fn: str, prompt: str = "", model: str = ""):
        self.fn         = fn
        self.prompt_len = len(prompt)
        self.model      = model
        self._start     = None
        self.response   = ""
        self.confidence = None
        self.retries    = 0
        self.error      = None

    def __enter__(self):
        self._start = time.time()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        duration_ms  = (time.time() - self._start) * 1000
        success      = exc_type is None
        if exc_val and not self.error:
            self.error = str(exc_val)
        record(
            fn=self.fn,
            duration_ms=duration_ms,
            prompt_len=self.prompt_len,
            response_len=len(self.response),
            success=success,
            model=self.model,
            confidence=self.confidence,
            error=self.error,
            retries=self.retries,
        )
        return False  # ne supprime pas l'exception


def load_traces() -> list:
    """Charge toutes les traces depuis logs/traces.jsonl."""
    if not os.path.exists(TRACE_FILE):
        return []
    traces = []
    with open(TRACE_FILE, encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if line:
                try:
                    traces.append(json.loads(line))
                except json.JSONDecodeError:
                    pass
    return traces


def clear_traces():
    """Vide le fichier de traces (nouveau run)."""
    if os.path.exists(TRACE_FILE):
        os.remove(TRACE_FILE)

File: perf_k6_qacart/scripts/agents/shared/tracer.js
'use strict';
// ============================================================
// Tracer — Instrumentation des appels LLM
// Écrit chaque appel dans logs/traces.jsonl (lu par observability-agent)
// ============================================================
const fs   = require('fs');
const path = require('path');

const FRAMEWORK  = path.join(__dirname, '..', '..', '..');
const LOGS_DIR   = path.join(FRAMEWORK, 'logs');
const TRACE_FILE = path.join(LOGS_DIR, 'traces.jsonl');

fs.mkdirSync(LOGS_DIR, { recursive: true });

function detectAgent() {
  const stack    = new Error().stack.split('\n');
  const agentsDir = path.normalize(path.join(__dirname, '..'));
  for (const line of stack) {
    const m = line.match(/\((.+\.js):\d+:\d+\)/);
    if (!m) continue;
    const fp = path.normalize(m[1]);
    if (fp.startsWith(agentsDir) && !fp.includes('llm.js') && !fp.includes('tracer.js')) {
      return path.basename(fp, '.js');
    }
  }
  return 'unknown';
}

function record({ fn, durationMs, promptLen, responseLen, success, model = '', confidence = null, error = null, retries = 0 }) {
  const entry = {
    ts:           new Date().toISOString(),
    agent:        detectAgent(),
    fn,
    model,
    duration_ms:  Math.round(durationMs * 10) / 10,
    prompt_len:   promptLen  || 0,
    response_len: responseLen || 0,
    success,
    retries,
  };
  if (confidence !== null) entry.confidence = Math.round(confidence * 1000) / 1000;
  if (error)               entry.error = String(error).slice(0, 200);
  try { fs.appendFileSync(TRACE_FILE, JSON.stringify(entry) + '\n', 'utf8'); } catch {}
}

// Span — mesure un bloc d'appel LLM
class Span {
  constructor(fn, prompt = '', model = '') {
    this.fn         = fn;
    this.promptLen  = typeof prompt === 'string' ? prompt.length : JSON.stringify(prompt).length;
    this.model      = model;
    this.response   = '';
    this.confidence = null;
    this.retries    = 0;
    this.error      = null;
    this._start     = null;
  }
  begin() { this._start = Date.now(); return this; }
  end(success = true) {
    record({
      fn:          this.fn,
      durationMs:  Date.now() - this._start,
      promptLen:   this.promptLen,
      responseLen: typeof this.response === 'string' ? this.response.length : JSON.stringify(this.response).length,
      success,
      model:       this.model,
      confidence:  this.confidence,
      error:       this.error,
      retries:     this.retries,
    });
  }
}

function loadTraces() {
  if (!fs.existsSync(TRACE_FILE)) return [];
  return fs.readFileSync(TRACE_FILE, 'utf8')
    .split('\n').filter(Boolean)
    .map(l => { try { return JSON.parse(l); } catch { return null; } })
    .filter(Boolean);
}

function clearTraces() {
  if (fs.existsSync(TRACE_FILE)) fs.unlinkSync(TRACE_FILE);
}

module.exports = { record, Span, loadTraces, clearTraces, TRACE_FILE };
